Improve tests and documentation for quad_form and quad_form_sym

mcol opened the pull request:
## Summary
This started out to be a simple clean up of the `quad_form` and `quad_form_sym` tests, but it turned into a rabbit hole. There were a few issues around those functions:
- the prim tests were in a single file: I've now split them into one file per function
- the mix tests for `quad_form_sym` were broken because their input was not symmetric: I now symmetrize the input matrix so that tests execute as expected (this also required fixing some tolerances)
- again in that test, variable `b22` was not initialized, and variable `b23` was not used
- there were no mix tests for `quad_form`: this is because there was no fwd version of that function, which I've now added
- some implementations could be consolidated by better template metaprogramming
- there was no doxygen documentation, and now all files have some.

Fixes #1695.

## Tests
Added and fixed as mentioned above.

## Side Effects

None.

## Checklist

- [X] Math issue #1695

- [X] Copyright holder: Marco Colombo

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

bob-carpenter commented on stan/math/fwd/fun/quad_form_sym.hpp:
```diff
@@ -8,40 +8,58 @@
   check_symmetric("quad_form_sym", "A", A);
-  Eigen::Matrix<fvar<T>, CB, CB> ret(multiply(transpose(B), multiply(A, B)));
+  Eigen::Matrix<T, CB, CB> ret(multiply(transpose(B), multiply(A, B)));
   return T(0.5) * (ret + transpose(ret));
```
[question]
Is there a reason to use the Eigen operators when the other `fvar` implementations seemed there to use the Stan functions like `multiply`?  The Stan functions should have a lower memory footprint.

Also using Stan functions, the 0.5 probably wouldn't need to be cast up to type `T`.

mcol replied:
I tried using multiply as suggested, but I couldn't get the tests to compile.

bob-carpenter replied:
OK---we can take that up in a separate issue.  It should be compilable.

bob-carpenter commented on stan/math/fwd/fun/quad_form_sym.hpp:
```diff
@@ -8,40 +8,58 @@
+                                           const Eigen::Matrix<TB, RB, 1>& B) {
   check_multiplicable("quad_form_sym", "A", A, "B", B);
   check_symmetric("quad_form_sym", "A", A);
   return dot_product(B, multiply(A, B));
```
[question]
Why isn't this one ensuring symmetry like the previous one?

mcol replied:
Because the result is a scalar.

bob-carpenter replied:
D'oh.

bob-carpenter commented on test/unit/math/prim/fun/quad_form_sym_test.cpp:
```diff
@@ -0,0 +1,74 @@
+#include <stan/math/prim.hpp>
+#include <gtest/gtest.h>
+
+TEST(MathMatrixPrim, quad_form_sym_mat) {
```
This should be testing boundary conditions of size 1 and size 0 matrices as well as these bigger cases as the other quad_form tests do.

mcol commented on stan/math/fwd/fun/quad_form.hpp:
```diff
@@ -0,0 +1,64 @@
+ * Compute the quadratic form B^T A B.
+ *
+ * @tparam Ta type of elements in the square matrix
+ * @tparam Ra number of rows in the square matrix, can be Eigen::Dynamic
```
I thought so. In reality we can't compile if we hardcode sizes because of `check_square`:
```
In file included from test/unit/math/prim/fun/quad_form_test.cpp:1:
In file included from ./stan/math/prim.hpp:10:
In file included from ./stan/math/prim/fun.hpp:238:
./stan/math/prim/fun/quad_form.hpp:31:3: error: no matching function for call to
      'check_square'
  check_square("quad_form", "A", A);
  ^~~~~~~~~~~~
test/unit/math/prim/fun/quad_form_test.cpp:22:31: note: in instantiation of
      function template specialization 'stan::math::quad_form<5, 5, 5, 5, double>'
      requested here
  matrix_d res2 = stan::math::quad_form(A, A);
                              ^
./stan/math/prim/err/check_square.hpp:21:13: note: candidate template ignored:
      could not match -1 against 5
inline void check_square(
            ^
1 error generated.
```

mcol commented:
I've addressed your comments as best as I could. In `test/unit/math/prim/fun/quad_form_sym_test.cpp` there was a duplicate test that I merged with the existing one.
